filter-kalman/ftnoir_filter_kalman.h

```diff
@@ -10,40 +10,115 @@
 
 #include "ui_ftnoir_kalman_filtercontrols.h"
 #include "opentrack/plugin-api.hpp"
-#include <opencv2/core.hpp>
-#include <opencv2/video/tracking.hpp>
-#include <vector>
+
+#include <atomic>
+#include <Eigen/Core>
+#include <Eigen/LU>
+
 #include <QString>
-#include <QElapsedTimer>
 #include <QWidget>
+
 #include "opentrack-compat/options.hpp"
 using namespace options;
 #include "opentrack-compat/timer.hpp"
 
+
+static constexpr int NUM_STATE_DOF = 12;
+static constexpr int NUM_MEASUREMENT_DOF = 6;
+using Matrix = Eigen::MatrixXd; // variable size, heap allocated, double valued
+// These vectors are compile time fixed size, stack allocated
+using StateVector = Eigen::Matrix<double, NUM_STATE_DOF, 1>;
+using PoseVector = Eigen::Matrix<double, NUM_MEASUREMENT_DOF, 1>;
+
+struct KalmanFilter
+{
+    Matrix
+        measurement_noise_cov,
+        process_noise_cov,
+        state_cov,
+        state_cov_prior,
+        kalman_gain,
+        transition_matrix,
+        measurement_matrix;
+    StateVector
+        state,
+        state_prior;
+    PoseVector
+        innovation;
+    void init();
+    void time_update();
+    void measurement_update(const PoseVector &measurement);
+};
+
+struct KalmanProcessNoiseScaler
+{
+    Matrix
+        innovation_cov_estimate,
+        base_cov; // baseline (unscaled) process noise covariance matrix
+    void init();
+    void update(KalmanFilter &kf, double dt);
+};
+
+
+struct DeadzoneFilter
+{
+    PoseVector
+        last_output,
+        dz_size;
+    DeadzoneFilter() :
+        last_output(PoseVector::Zero()),
+        dz_size(PoseVector::Zero())
+    {}
+    void reset() {
+        last_output = PoseVector::Zero();
+    }
+    PoseVector filter(const PoseVector &input);
+};
+
+
 struct settings : opts {
-    value<int> noise_stddev_slider;
-    // slider for noise_stddev goes 0->(mult_noise_stddev * 100)
-    static constexpr double mult_noise_stddev = .5;
-    settings() : opts("kalman-filter"), noise_stddev_slider(b, "noise-stddev", 40)
+    value<int> noise_rot_slider_value;
+    value<int> noise_pos_slider_value;
+
+    static constexpr double adaptivity_window_length = 0.5; // seconds
+    static constexpr double deadzone_scale = 2.;
+    static constexpr double deadzone_exponent = 4.0;
+    // these values worked best for me (MW) when taken with acompanying measured noise stddev of ca 0.1 (rot) and 0.01 (pos). 
+    static constexpr double process_sigma_pos = 0.05;
+    static constexpr double process_simga_rot = 0.5;
+
+    static double map_slider_value(int v)
+    {
+        return std::pow(10., v * 0.04 - 3.);
```

Comment:
This scaling's weird from my usage. Can we please make it linear? The ui slider can have larger resolution with higher single/page step value.